### ISSUE TITLE
Resolving Confusions around Aliases

### ISSUE THREAD

**Issue by vhmth** [tshaddix/react-chrome-redux#46]
@tshaddix right now our biggest point of confusion for most folks is the separation of concerns around what an alias and an action is. This introduces a new paradigm where you have to mentally remember that aliases run in one environment and actions in another. I think it largely confuses newbie Chrome extension devs as well.

My proposal is that we figure out a way to ensure that *all* actions get run/dispatched on the background. I'm not even sure if there's a clean way to ensure this, but I think opening the conversation is worthwhile.

**Comment by tshaddix** [comment]
@vhmth You're right. It's definitely not transparent what's going on under the hood with aliases.

I'm not convinced I've done a good enough job documenting aliases when I wrote the documentation. I'd love to see if putting a little more time into the docs on my end could solve this.

**Comment by AgrimPrasad** [comment]
@tshaddix @vhmth I seem to be getting confused too about aliases... I'm trying to get the active tab from the chrome.tabs api in the alias, but seems like I'm unable to do so. Could someone explain why?

On the actual popup UI for the following code, I see:

> Current Host: Unknown Host

... instead of the hostname of the currently active tab.

aliases.js:
```

import actionTypes from '../actions/actionTypes';
import URL from 'url-parse';

const selectHost = () => {
	let host = 'Unknown Host';
	chrome.tabs.query({currentWindow: true, active: true}, (tabs) => {
		host = new URL(tabs[0].url).hostname;
    });

    const action = {
		type: actionTypes.SELECT_HOST,
		host
	};
	return action;
};

export default {
	[actionTypes.SELECT_HOST]: selectHost
};

```

event/index.js:

```
import { createStore, applyMiddleware } from 'redux';
import { alias, wrapStore } from 'react-chrome-redux';
import thunkMiddleware from 'redux-thunk';
import createLogger from 'redux-logger';

import rootReducer from '../../reducers';
import actionTypes from '../../actions/actionTypes';
import aliases from '../../aliases/aliases';

const loggerMiddleware = createLogger();
const preloadedState = {};
const enhancer = applyMiddleware(
	thunkMiddleware,
	loggerMiddleware,
	alias(aliases)
);
const store = createStore(rootReducer, 
	preloadedState,
	enhancer);

wrapStore(store, {
  portName: 'appPort'
});

```

popup/App.jsx:

```

import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actionCreators from '../../actions/actionCreators';

class App extends Component {
  componentDidMount() {
    this.props.selectHost();
  }

  render() {
    return (
        <div>
          <h2>
            Current Host: {this.props.selectedHost}
          </h2>
        </div>
      );
  }
}

App.propTypes = {
  dispatch: React.PropTypes.func,
  selectedHost: React.PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  selectedHost: state.selectedHost
});

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(App);

```

actionCreators.js (dummy action):

```
export function selectHost() {
    const action = {
		type: actionTypes.SELECT_HOST,
	};
	return action;
}

```

selectedHost.js (reducer):

```

import actionTypes from '../actions/actionTypes';

const initialState = 'yahoo.com';

export default (state = initialState, action) => {
	switch (action.type) {
		case actionTypes.SELECT_HOST:
			return action.host;
		default:
			return state;
	}
}

```

**Comment by vhmth** [comment]
@AgrimPrasad have you inspected what the current active tab is by break-pointing? Is it possibly picking up on some tab context the popup window owns?

**Comment by AgrimPrasad** [comment]
@vhmth Thanks for responding. I've tried to set breakpoints in Chrome, and added a console.log inside the alias, but curiously, the breakpoint is not hit + nothing is logged to the console. I've also tried to use the debugger; keyword as per https://developers.google.com/web/tools/chrome-devtools/javascript/step-code without success... Nonetheless, the state object "host" variable is set to 'alias Host' which means that the alias code is indeed being executed, but the breakpoint is not set.

aliases.js:

```
import actionTypes from '../actions/actionTypes';
import URL from 'url-parse';

const selectHost = () => {
	debugger;
	console.log('In selectHost() alias');
	let host = 'alias Host';
	chrome.tabs.query({currentWindow: true, active: true}, (tabs) => {
		host = new URL(tabs[0].url).hostname;
    });

    const action = {
		type: actionTypes.SELECT_HOST,
		host
	};
	return action;
};

export default {
	[actionTypes.SELECT_HOST]: selectHost
};
```
Screenshot which shows that the "alias host" from above does get added to the state object.

![image](https://cloud.githubusercontent.com/assets/6386887/22646466/82cf67d8-eca7-11e6-87f2-da9625e5cb9b.png)

I'm quite lost here... Could someone please take a look at my github repo to see what might be happening? Any help would be really appreciated as I've been stuck for 3 days with this and wondering if it's better to not use aliases at all.

https://github.com/AgrimPrasad/mozilla-observatory-extensions/tree/aliasExperiment

**Comment by vhmth** [comment]
Hey @AgrimPrasad let me know if this makes sense:

https://www.useloom.com/share/796ab600ecae11e6904fed68ddabdd65

**Comment by AgrimPrasad** [comment]
Thanks a lot @vhmth! Yes, that cleared up a lot of things for me: aliases, async actions in general and how to debug background pages. 

Some notes for other new chrome extensions/redux developers who might be having issues:

1. Alias is basically an action in the context of the background page which is mapped to an action in the context of the popup/content script page. 
2. If you use a Middleware such as Redux thunk and follow the official docs, make sure that all the npm packages are updated. I was stuck for several hours making async actions work, merely because the official docs use a newer Redux syntax whereas my Redux version was old. @tshaddix I have opened an issue regarding this in https://github.com/tshaddix/react-chrome-redux examples which I have used as a minimal boilerplate. 
3. The background page inspector is different from the popup inspector and accessible from the chrome extensions page as mentioned in http://stackoverflow.com/questions/10081898/how-to-debug-google-chrome-background-script
4. chrome.* api which are specific to Chrome use callbacks, whereas browser.* api from [WebExtensions](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Chrome_incompatibilities) use Promises. The Promises based api can be used in Chrome also by using Mozilla's [browser-polyfill](https://github.com/mozilla/webextension-polyfill)
